### PR TITLE
`resplit` with Custom MPI Datatypes and AlltoAllW

### PR DESCRIPTION
**Pull request by JuanPedroGHM**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description
Rewritten most of `resplit` to use `Alltoallw` and custom data types. 

Issue/s resolved: #

## Changes proposed:
- MPICommunicator
   -  `Alltoallw` operation, that mimics the MPI `Alltoallw` interface.
  - `mpi_type_of ` class method for easy of use
  - `_create_recursive_vector` to handle subarray datatype creation for non-contiguous send buffers. 
- Manipulations
  - `_axis2axis` method to handle all non-trivial replits
- Tiling
  - `get_subarray_params` method to calculate MPI subarray type params

## Type of change
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Memory requirements
Coming soon...

## Performance
Coming soon...

#### Does this change modify the behaviour of other functions? If so, which?
yes

Probably most high level ones.

### CODE REVIEW

**Comment by JuanPedroGHM** [comment]
Some results from Horeka using the following code: 

```python
from mpi4py import MPI
import heat as ht
import argparse
import perun
import torch

from heat.core.communication import CUDA_AWARE_MPI
print(f"CUDA_AWARE_MPI: {CUDA_AWARE_MPI}")

@perun.monitor()
def cpu_contiguous(a):
    a = a.resplit(4)
    a.resplit_(3)


@perun.monitor()
def cpu_noncontiguous(a):
    a = a.resplit(0)
    a.resplit_(2)


@perun.monitor()
def gpu_contiguous(a):
    a = a.resplit(4)
    a.resplit_(3)

@perun.monitor()
def gpu_noncontiguous(a):
    a = a.resplit(0)
    a.resplit_(2)

if __name__ == "__main__":
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()
    
    # Contiguous data creation
    shape = [100, 50, 50, 20, 250]
    n_elements = ht.array(shape).prod().item()
    Mem = n_elements * 8 / 1e9
    base_array = torch.arange(0, n_elements, dtype=torch.float64).reshape(shape) * (rank+1)
    
    print(f"Rank {rank} - Local Shape: {shape} - Memory: {Mem * size} GB - Per rank: {Mem} GB")
    
    # CPU contiguous data
    print("CPU contiguous data")
    a = ht.array(base_array, dtype=ht.float64, is_split=0, copy=True)
    print(f"Rank {rank} - Shape: {a.shape} - Split: {a.split} - Lshape: {a.lshape} - Device: {a.device}")
    cpu_contiguous(a)
    del a
    
    # CPU non-contiguous data
    print("CPU non-contiguous data")
    a = ht.array(base_array, dtype=ht.float64, is_split=1, copy=True).transpose((1,0,4,3,2))
    print(f"Rank {rank} - Shape: {a.shape} - Split: {a.split} - Lshape: {a.lshape} - Device: {a.device}")
    cpu_noncontiguous(a)
    del a

    # GPU contiguous data
    print("GPU contiguous data")
    a = ht.array(base_array, dtype=ht.float64, device="cuda", is_split=0, copy=True)
    print(f"Rank {rank} - Shape: {a.shape} - Split: {a.split} - Lshape: {a.lshape} - Device: {a.device}")
    gpu_contiguous(a)
    del a
    
    # GPU non-contiguous data
    print("GPU non-contiguous data")
    a = ht.array(base_array, dtype=ht.float64, device="cuda", is_split=1, copy=True).transpose((1,0,4,3,2))
    print(f"Rank {rank} - Shape: {a.shape} - Split: {a.split} - Lshape: {a.lshape} - Device: {a.device}")
    gpu_noncontiguous(a)
    del a

    torch.cuda.empty_cache()
```
## Runtime
image

## Memory
image

## GPU Memory
image

**Comment by ClaudiaComito** [comment]
And the runtime axis is logarithmic? :grin:  Fantastic @JuanPedroGHM !

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
@JuanPedroGHM I will review next week, but in the meantime - was `resplit` ever tested on column-major arrays? To me it looks like we forgot that one. If so, would you add a test for DNDarrays with order="F"? There are some examples in `test_dndarray.test_stride_and_strides`.

Thanks again for the fantastic job!

File: heat/core/dndarray.py [hunk 1]
```diff
@@ -1471,68 +1471,25 @@ def resplit_(self, axis: int = None):
...
+        arr_tiles = tiling.SplitTiles(self)
+        new_tiles = tiling.SplitTiles(self)
+
+        gshape = self.shape
+        new_lshape = list(gshape)
+        new_lshape[axis] = int(arr_tiles.tile_dimensions[axis][self.comm.rank].item())
+
+        recv_buffer = torch.empty(
+            tuple(new_lshape), dtype=self.dtype.torch_type(), device=self.device.torch_device
         )
```
Comment (ClaudiaComito) [inline heat/core/dndarray.py:1483]: Is this still in-place? 😬 Not sure this function ever was in place, really. Expand docs?

What's the  difference in memory usage between `array.resplit_(newaxis)`  and `array = ht.resplit(array, newaxis)`?

File: heat/core/manipulations.py [hunk 2]
```diff
@@ -3537,34 +3537,15 @@ def resplit(arr: DNDarray, axis: int = None) -> DNDarray:
             gathered, is_split=axis, device=arr.device, comm=arr.comm, dtype=arr.dtype
         )
         return new_arr
-    arr_tiles = tiling.SplitTiles(arr)
+
```
Comment (ClaudiaComito) [inline heat/core/manipulations.py:3540]: Should we update the docstring here, and mention the Dalcin paper?

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
@JuanPedroGHM thank you so much for this. I have a few comments, mostly from the point of view of maintainability. Great job!

File: heat/core/communication.py [hunk 3]
```diff
@@ -1443,6 +1459,181 @@ def Alltoallv(
...
+    ):
+        """
+        Generalized All-to-All communication allowing different counts, displacements and datatypes for each partner.
+
+        Parameters
+        ----------
+        sendbuf: Union[DNDarray, torch.Tensor, Any]
+            Buffer address of the send message
+        recvbuf: Union[DNDarray, torch.Tensor, Any]
+            Buffer address where to store the result
```
[the remark this answers is not in the document]
Comment (JuanPedroGHM) [reply heat/core/communication.py:1475]: That is definitely a big problem in the whole problem in the whole communication.py file. I guess it was made like this originally to support ```_alltoall_like``` and similar methods that need a flexible function signature. 

For me, what we need is to define a more consistent communication interface, for example, one where all the communication buffers are a tuple of a torch array, and a collection of views from that buffer that we use to define the data types. What do you think?

For now, I'll expand on the actual contents of the buffer on the doc string.

File: heat/core/communication.py [hunk 4]
```diff
@@ -1443,6 +1459,181 @@ def Alltoallv(
...
+            Index of the first element of the subarray in the original array.
+
+        Returns
+        -------
+        MPI.Datatype
+            The created recursive vector datatype.
+
+        Notes
+        -----
+        This function creates a recursive vector datatype by defining vectors out of the previous datatype with specified strides and sizes. The extent of the new datatype is set to the extent of the basic datatype to allow interweaving of data.
```
[the remark this answers is not in the document]
Comment (JuanPedroGHM) [reply heat/core/communication.py:1583]: It is a weird name for the length, or how many bits, the data type uses. Every MPI datatype has a symbolic extent, that we use to read non-contiguous data in the right order, and a real extent. More here: https://enccs.github.io/intermediate-mpi/derived-datatypes-pt2/

File: heat/core/dndarray.py (hunk 1 above)
Comment (JuanPedroGHM) [reply heat/core/dndarray.py:1483]: Well, it is the type of in place resplit as before, which is not really in place for on the ```larray```.  The DNDArray is reused and all the other properties are rewritten, but a new ```larray``` is created. 

We could keep the same larray if the incoming and out coming data has the same length, but it will have to be written in a non-contiguous way, which is not impossible, but not ideal in my opinion.

File: heat/core/manipulations.py (hunk 2 above)
Comment (JuanPedroGHM) [reply heat/core/manipulations.py:3540]: Do you mean the FFT paper? I would reference it inside ```Alltoallw```, as much of the relevant code landed there, or in ```tiling.py```

File: heat/core/tests/test_dndarray.py [hunk 5]
```diff
@@ -1106,6 +1106,33 @@ def test_resplit(self):
...
+        # 3D non-contiguous resplit testing (Column mayor ordering)
+        torch_array = torch.arange(100, device=self.device.torch_device).reshape((10, 5, 2))
+        print(torch_array.device)
+        heat_array = ht.array(torch_array, split=2, order="F")
+        print(heat_array.device)
+        heat_array.resplit_(axis=1)
+        print(heat_array.device)
+        res = np.arange(100).reshape(10, 5, 2)
+        print(heat_array.device)
+        print(ht.array(res).device)
```
[the remark this answers is not in the document]
Comment (JuanPedroGHM) [reply heat/core/tests/test_dndarray.py:1118]: They only appear when something is wrong, so I forgot to take them out.

File: heat/core/manipulations.py (hunk 2 above)
Comment (ClaudiaComito) [reply heat/core/manipulations.py:3540]: OK for referencing inside Alltoallw

File: heat/core/manipulations.py [hunk 6]
```diff
@@ -3575,6 +3556,67 @@ def resplit(arr: DNDarray, axis: int = None) -> DNDarray:
...
 
+def _axis2axisResplit(
+    comm: Communication,
+    source_axis: int,
+    source_array: torch.Tensor,
+    source_tiles: tiling.SplitTiles,
+    target_axis: int,
+    target_array: torch.Tensor,
+    target_tiles: tiling.SplitTiles,
+) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/manipulations.py]: Were the arguments swapped in the end? 🤔

[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/manipulations.py]: OK apologies, I see now you weren't done with this PR yet. No hurry, I'm away this week.
